src/toil/__init__.py

```diff
@@ -456,215 +453,229 @@ def str_to_datetime(s):
         """
         return datetime.strptime(s, datetime_format)
 
-    class BotoCredentialAdapter(provider.Provider):
-        """
-        Adapter to allow Boto 2 to use AWS credentials obtained via Boto 3's
-        credential finding logic. This allows for automatic role assumption
-        respecting the Boto 3 config files, even when parts of the app still use
-        Boto 2.
+    # Now we have defined the adapter class. Patch the Boto module so it replaces the default Provider when Boto makes Providers.
+    provider.Provider = BotoCredentialAdapter
+
+try:
+    from boto import provider
+    base_class = provider.Provider
+except ImportError:
+    base_class = object
+
+
+class BotoCredentialAdapter(base_class):
+    """
+    Adapter to allow Boto 2 to use AWS credentials obtained via Boto 3's
+    credential finding logic. This allows for automatic role assumption
+    respecting the Boto 3 config files, even when parts of the app still use
+    Boto 2.
+    
+    This class also handles cacheing credentials in multi-process environments
+    to avoid loads of processes swamping the EC2 metadata service.
+    """
+    
+    from botocore.session import Session
+    from botocore.credentials import create_credential_resolver, RefreshableCredentials, JSONFileCache
```

Comment:
You might want to move these imports into the `try` above. I don't think they'll work here.